src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -423,5 +657,95 @@ private boolean canAccess(Authentication user, SecureTreeNode node) {
             return access;
         }
     }
+    
+    private SecureTreeNode getNodeForGroup(LayerGroupInfo lg) {
+        SecureTreeNode node;
+        if(lg.getWorkspace() == null) {
+            node = root.getNode(lg.getName());
+        } else {
+            String[] path = getLayerGroupPath(lg);
+            node = root.getNode(path);
+        }
+        return node;
+    }
+    
+    /**
+     * This listener keeps the "layer group" flags in the authorization tree current, in order
+     * to optimize the application of layer group containment rules
+     */
+    class CatalogChangeListener implements CatalogListener {
+
+        @Override
+        public void handleAddEvent(CatalogAddEvent event) throws CatalogException {
+            if(event.getSource() instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                addGroupInfo(lg);
+            } 
+        }
+
+        private void addGroupInfo(LayerGroupInfo lg) {
+            SecureTreeNode node = getNodeForGroup(lg);
+            if(node != null) {
+                updateContainedLayerIds(node, lg);
+            }
+        }
+
+        @Override
+        public void handleRemoveEvent(CatalogRemoveEvent event) throws CatalogException {
+            if(event.getSource() instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                clearGroupInfo(lg);
+            }
+        }
+
+        @Override
+        public void handleModifyEvent(CatalogModifyEvent event) throws CatalogException {
+            final CatalogInfo source = event.getSource();
+            if(source instanceof LayerGroupInfo) {
+                LayerGroupInfo lg = (LayerGroupInfo) event.getSource();
+                // was the layer group renamed or moved?
+                int nameIdx = event.getPropertyNames().indexOf("name");
+                int wsIdx = event.getPropertyNames().indexOf("workspace");
+                if(nameIdx != -1 || wsIdx != -1) {
+                    // group has been modified, clear the old node (the new one will be updated in ppost modify)
+                    clearGroupInfo(lg);
```

Comment:
Wait, the rule is still there and that's normal, and "your fault", but the security subsystem should stop tracking the  catalog ids contained in that rule because it's now dangling. Is that not happening? I?ll check.